mlflow/crewai/__init__.py

```diff
@@ -0,0 +1,169 @@
+"""
+The ``mlflow.crewai`` module provides an API for tracing CrewAI AI agents.
+"""
+
+from mlflow.crewai.autolog import (
+    patched_class_call,
+)
+from mlflow.utils.annotations import experimental
+from mlflow.utils.autologging_utils import autologging_integration, safe_patch
+
+FLAVOR_NAME = "crewai"
+
+
+@experimental
+@autologging_integration(FLAVOR_NAME)
+def autolog(
+    log_traces: bool = True,
+    disable: bool = False,
+    silent: bool = False,
+):
+    """
+    Enables (or disables) and configures autologging from CrewAI to MLflow.
+    Only synchronous calls are supported. Asynchnorous APIs and streaming are not recorded.
+
+    Args:
+        log_traces: If ``True``, traces are logged for CrewAI agents.
+            If ``False``, no traces are collected during inference. Default to ``True``.
+        disable: If ``True``, disables the CrewAI autologging. Default to ``False``.
+        silent: If ``True``, suppress all event logs and warnings from MLflow during CrewAI
+            autologging. If ``False``, show all events and warnings.
+    """
+    # TODO: handle asynchronous tasks and crew executions
+    # TODO: interface of tool in CrewAI is changing drastically. Add patching once it's stabilized
+    import crewai
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "kickoff",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "kickoff_for_each",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Crew,
+        "train",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Agent,
+        "execute_task",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Task,
+        "execute_sync",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.LLM,
+        "call",
+        patched_class_call,
+    )
+
+    safe_patch(
+        FLAVOR_NAME,
+        crewai.Flow,
+        "kickoff",
+        patched_class_call,
+    )
```

Comment:
As discussed in the design doc, let's patch `BaseTool` class to capture tool execution information.

Reply:
As I added the comment, the BaseTool class is actually not used by CrewAI. Paching BaseTool.run does not work.

Reply:
It is used tho  https://docs.crewai.com/concepts/tools

Reply:
btw for built-in tools are defined in https://github.com/crewAIInc/crewAI-tools not the core repo, which implement BaseTool

Reply:
I mean, the function itself is not used directly, inside the library which prevent patching. Also the implementation of Tool is changing, even the latest code is very different from the master branch. I'd wait until having more stability.

Reply:
I did a bit of research:
* In version `0.83`, they are using LangChain's tool implementation and call `tool.invoke` within the agent. https://github.com/crewAIInc/crewAI/blob/v0.83.0/src/crewai/agent.py#L402
* In the latest `main` branch, they've added `CrewStructuredTool` to remove LangChain dependency.

The new structured tool class will be a standard for tool calling. I agree the code base is changing rapidly right now, but we must have tool spans as well, otherwise users cannot debug/measure tool execution. Can we leave a TODO comment and file a time-bounded JIRA?

Reply:
I've added a comment [here](https://github.com/mlflow/mlflow/pull/13903/files/2c9d3d3f7729f425e841b5e662091c4a0756bbf9#diff-c754f6b0a60b7373d437491941f830ce9a6d616e26bd39366b0d984b1d3808b1R33), let me create a ticket for it too.